docs/design/reserve-action.md

```diff
@@ -0,0 +1,278 @@
+were bound to the node and begin to run. The `pod-4` has the highest priority in the unscheduled pod queue and will be 
+prioritized to be schedule. In `Time2`, `pod-1` was completed, and `1C1G` resource was released from node. The scheduler
+try to schedule `pod-4`, but idle resource of node can not satisfy pod's requirement. The `pod-4` can not be scheduled and 
+was skipped. The schedule try to schedule `pod-5` and succeed in `Time3`. In `Time4`, `pod-2` and `pod-3` were `completed`
+and `2C2G` was released from node. Because of that the idle resource of node can not satisfy requirement of pod `pod-4`, `pod-4`
+was skipped. The `pod-6` and `pod-7` were bound to node and begin to run. In `Time6` all pods except pod `pod-4` were completed
+and all resources of node were release. The scheduling condition of pod `pod-4` is satisfied, `pod-4` was bound to node and 
+begin to run. In the above scenario, the `pod-4` can not be scheduled in time because a large number of small resource requirement
+pods occupied the node resources.
+ 
+In order to solve such problems, the reserve action is introduced into the scheduling policy. The `reserve` action can 
+prevents the large resource requirement task from being starved by reserving node resources for it and not allowing the other 
+tasks to be scheduled on the node.
+
+![reserve-resource-for-pod](../../docs/images/reserve-resource-for-pod.png)
+
+Different from the original scheduling logic, in `Time4`, although there are `2C2G` idle resources on the node, it does not
+schedule new tasks to the nodes, but reserves these resources for `pod-4`. In `Time5` `pod-6` and `pod-7` remain `Pending`. 
+In `Time 6`, `pod-5` completed and all resources on node were released. The scheduling condition of `pod-4` is satisfied, `pod-4`
+is bound to the node and starts `Running`.
+
+## Function Detail 
+
+### Terminology
+
+#### Job Ready
+
+If the running condition of a job is satisfied, then the job is in the `Job Ready` state, and after scheduling, the job can be in the 
+`Job Running` state. The running condition of a job may include restrictions such as `gang-scheduler` and so on.
+
+#### Job Condition Ready
+
+If the running condition of a job is satisfied under certain conditions, then the state of the job can be called `Conditional Running`.
+When after the node reserving resources for the tasks under the job, the job can reach the running condition, then the state of 
+the job can be called `Condition Ready`.
+
+#### Job starving
+
+The starving job needs to meet the following conditions
+
+* The creation time of job is greater than the set starving job threshold time
+* Job state is not `Job Ready`
+* Job state is not `Job Condition Ready`
+
+If a job has been created for a long time, but it is still not `running`, it is at risk of starving. If nodes under the cluster
+have reserved resource for the starving job, then there is no need to reserve resources for it again.
+
+#### ForceAllocated
```

Comment:
Why introduces two state? I'm thinking just trace `Reserved` resources should be enough, ignore the relationship between idle resources and request.

Reply:
The introduction of the `ForceAllocated` state is expected to replace the `allocate` state. In the `reserve` phase, for starving jobs, if node idle resources satisfy pod's resource requirement, the pod scheduling may be completed. So, the pod needs an identifier that occupies the actual resource. In order to avoid preempting the resources that have been allocated to the pod in the `backfill` phase, a new state is defined. Introducing the `FakeAllocate` state is to make an identifier for whether the node reserves resources and whether the task is reserved. If task with `FakeAllocate` status is added to the node, idle resource of node will be decreased, so the node with `FakeAllocate` tasks indicates that the idle resources of the node are tight and need to overdraw resources. The `Reserved` resource can also determine whether the node has reserved resources for the task, and whether there are remaining resources to schedule other tasks, but it is possible that this judgment is not as clear as the status identifier.

Reply:
A new `Reserved` state is added to indicate that the task is reserved on the node, and the `ForceAllocate` state is replaced by the `Allocated` state.